sdk/keyvault/Azure.Security.KeyVault.Secrets/tests/ChallengeBasedAuthenticationPolicyTests.cs

```diff
@@ -0,0 +1,233 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Diagnostics.Tracing;
+using System.IO;
+using System.Text;
+using System.Text.Json;
+using System.Text.RegularExpressions;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Core;
+using Azure.Core.Diagnostics;
+using Azure.Core.Pipeline;
+using Azure.Core.Testing;
+using Azure.Security.KeyVault.Tests;
+using NUnit.Framework;
+
+namespace Azure.Security.KeyVault.Secrets.Tests
+{
+    public class ChallengeBasedAuthenticationPolicyTests
+    {
+        private const string TenantId = "72f988bf-86f1-41af-91ab-2d7cd011db47";
+        private const string AccessToken = "NzJmOTg4YmYtODZmMS00MWFmLTkxYWItMmQ3Y2QwMTFkYjQ3"; // base64(TenantId)
+        private const string VaultHost = "test.vault.azure.net";
+        private static readonly Uri s_vaultUri = new Uri("https://" + VaultHost);
+
+        [Test]
+        public async Task SingleRequest()
+        {
+            MockTransport transport = new MockTransportBuilder().Build();
+            SecretClientOptions options = new SecretClientOptions
+            {
+                Transport = transport,
+                Diagnostics =
+                {
+                    LoggedHeaderNames = { "*" },
+                    IsLoggingContentEnabled = true,
+                },
+            };
+
+            using AzureEventSourceListener logger = AzureEventSourceListener.CreateTraceLogger(EventLevel.Verbose);
+            SecretClient client = new SecretClient(s_vaultUri, new MockCredential(transport), options);
+
+            KeyVaultSecret secret = await client.GetSecretAsync("test-secret").ConfigureAwait(false);
+            Assert.AreEqual("secret-value", secret.Value);
+        }
+
+        [TestCase(10, 0, 0)]
+        [TestCase(10, 0, 500)]
+        public async Task MultipleRequests(int numberOfRequests, int minDelay, int maxDelay)
+        {
+            Random rand = new Random();
+
+            MockTransportBuilder builder = new MockTransportBuilder();
+            builder.Request += (sender, args) =>
+            {
+                int delay;
+                lock (rand)
+                {
+                    delay = rand.Next(minDelay, maxDelay);
+                }
+
+                Thread.Sleep(delay);
+            };
+
+            MockTransport transport = builder.Build();
+            SecretClientOptions options = new SecretClientOptions
+            {
+                Transport = transport,
+                Diagnostics =
+                {
+                    LoggedHeaderNames = { "*" },
+                    IsLoggingContentEnabled = true,
+                },
+            };
+
+            using AzureEventSourceListener logger = AzureEventSourceListener.CreateTraceLogger(EventLevel.Verbose);
+            SecretClient client = new SecretClient(s_vaultUri, new MockCredential(transport), options);
```

Comment:
Do we need to go though SecretClient here? Can we write a test that only uses policy in question?

Reply:
It's possible, but why write more code? I'd rather keep the tests as close to production code as possible, and this is the most straight-forward way. Unless you feel strongly, I'd rather keep this as-is.

Reply:
I would expect that testing only that policy would involve less code that going through the `SecretClient`.

Usually I try to keep tests that verify race conditions very targeted because they tend to be pretty sensitive and the more moving parts there are the harder it gets to keep the tests consistently reproducing the issue.